src/Components/Server/src/Circuits/RemoteRenderer.cs

```diff
@@ -203,28 +204,62 @@ public void OnRenderCompleted(long incomingBatchId, string errorMessageOrNull)
+            // We send batches right away, and queue them until we receive an ACK.
+            // If one or more client ACKs get lost we might receive an ack for a higher batch.
+            // We confirm all previous batches at that point (because receiving an ack is guarantee
+            // from the client that it has received and successfully applied all batches up to that point).
+
+            // If receive an ack for a previously rendered batch, its an error, as the messages are
+            // guranteed to be delivered in order, so a message for a render batch of 2 will never arrive
+            // after a message for a render batch for 3.
+            // If that were to be the case, it would just be enough to relax the checks here and simply skip
+            // the message.
+
+            // A batch might get lost when we send it to the client. If the client receives a newer batch when
+            // it was expecting a previous batch, it can simply force a reconnection.
+
             // Even though we're not on the renderer sync context here, it's safe to assume ordered execution of the following
             // line (i.e., matching the order in which we received batch completion messages) based on the fact that SignalR
             // synchronizes calls to hub methods. That is, it won't issue more than one call to this method from the same hub
             // at the same time on different threads.
-            if (!PendingRenderBatches.TryDequeue(out var entry) || entry.BatchId != incomingBatchId)
+            if (!PendingRenderBatches.TryPeek(out var entry) || incomingBatchId < entry.BatchId)
             {
                 HandleException(
                     new InvalidOperationException($"Received a notification for a rendered batch when not expecting it. Batch id '{incomingBatchId}'."));
             }
             else
             {
-                if (errorMessageOrNull == null)
+                var lastBatchId = entry.BatchId;
+                // Order is important here so that we don't prematurely dequeue the last entry
+                while (PendingRenderBatches.TryPeek(out entry) && entry.BatchId <= incomingBatchId)
                 {
-                    Log.CompletingBatchWithoutError(_logger, entry.BatchId);
+                    lastBatchId = entry.BatchId;
+                    PendingRenderBatches.TryDequeue(out _);
+                    ProcessPendingBatch(errorMessageOrNull, entry);
                 }
-                else
+
+                if (lastBatchId < incomingBatchId)
                 {
-                    Log.CompletingBatchWithError(_logger, entry.BatchId, errorMessageOrNull);
+                    Log.ReceivedUnexpectedBatchId(_logger, incomingBatchId, lastBatchId);
+                    HandleException(
+                        new InvalidOperationException($"Received an acknowledgement for batch with id '{incomingBatchId}' when the last batch produced was '{lastBatchId}'."));
```

Comment:
Also should be logged as debug, I think

Reply:
In fact, is it enough to let `HandleException` log it? This will give us two log entries, won't it?

Reply:
Yep, but we want to terminate the circuit, and I want to do that change as part of the warning cleanups change. We might be able to get rid of the log statement.